src/Functions/namedTupleItems.cpp

```diff
@@ -0,0 +1,130 @@
+#include <Functions/IFunction.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeString.h>
+#include <Columns/ColumnTuple.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnsNumber.h>
+#include <Common/assert_cast.h>
+#include <memory>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_INDEX;
+}
+
+namespace
+{
+
+/** Extract element of tuple by constant index or name. The operation is essentially free.
+  * Also the function looks through Arrays: you can get Array of tuple elements from Array of Tuples.
+  */
+class FunctionNamedTupleItems : public IFunction
+{
+public:
+    static constexpr auto name = "namedTupleItems";
```

Comment:
After discussion with @alexey-milovidov we decided that better name is tupleToNameValuePairs.